flaml/automl/model.py

```diff
@@ -1567,13 +1567,6 @@ def fit(self, X_train, y_train, budget=None, free_mem_ratio=0, **kwargs):
                 # for xgboost>=1.6.0, pop callbacks to enable pickle
                 callbacks = self.params.pop("callbacks")
                 self._model.set_params(callbacks=callbacks[:-1])
-            best_iteration = (
-                getattr(self._model.get_booster(), "best_iteration", None)
-                if isinstance(self, XGBoostSklearnEstimator)
-                else self._model.best_iteration_
-            )
-            if best_iteration is not None:
-                self._model.set_params(n_estimators=best_iteration + 1)

```

Comment:
Hi @thinkall, no worries - best to be cautious here. 

The code highlighted above doesn't overwrite the `LGBMEstimator` `n_estimators` value - it overwrites the underlying `LGBMClassifier` or `LGBMRegressor` value. Please note that this code is also run **after** the `._fit()` has been called, so the highlighted code doesn't actually affect the actual fitting of the model, it just affects the configuration of `automl.model.model` i.e. the underlying model returned to the user.

If I run `test_reproducibility_of_underlying_classification_models[lgbm]` (which is commented out on FLAML main), I get the below error, showing that the FLAML best loss cannot be reproduced using the underlying `LGBMClassifier` model. It _is_ however reproducible using the FLAMLised LGBMEstimator.
![image](https://github.com/user-attachments/assets/982c0254-e5b6-4fc9-8a29-bd0b63f89ff0)

Please note that for the above screenshot, I'm running this test on the FLAML main branch, with the only difference being that I uncommented this test out (plus a few print statements).

Whilst you are accurate to say that `automl.best_config` is correct, that's not the issue here. `automl.best_config` returns the configuration of the best FLAMLised estimator, in this case, the `LGBMEstimator` configuration. Specifically in this failing test, I am trying to reproduce the FLAML best loss using the underlying `LGBMClassifier` or `LGBMRegressor` (not the FLAMLised estimator - that's fine). 

By investigating the configuration of these underlying models, I can see that every underlying `LGBMClassifier` or `LGBMRegressor` returned by FLAML has `n_estimators=1`. This is because `self._best_iteration_` always returns 0 for these estimators. The configuration of the underlying model is being overwritten _after_ it has been fitted, which doesn't make sense. I can't see why the `n_estimators` values should differ between the `LGBMEstimator` and underlying `LGBMClassifier`.   

If instead of logging the LGBMEstimator configs, we logged the LGBMClassifier configs, you would see that `n_estimators` is always 1 - regardless of the `n_estimators` value which was actually used. In fact, I tried exactly that! I added print statements to the above code i.e.:

```
            best_iteration = (
                getattr(self._model.get_booster(), "best_iteration", None)
                if isinstance(self, XGBoostSklearnEstimator)
                else self._model.best_iteration_
            )
            print('previous n_estimators value', self._model.get_params()['n_estimators'])
            print('best_iteration', best_iteration)
            if best_iteration is not None:
                self._model.set_params(n_estimators=best_iteration + 1)
```

and the following was printed to my console:
![image](https://github.com/user-attachments/assets/cdef8586-91c5-46ed-994d-b59c16a538ec)

`best_iteration` is **always** 0, meaning that we **always"" assign the underlying model `n_estimators=1`

I am aware that other estimators depends on this. I've tested XGBoost specifically, and in every test I ran, `best_estimators` was identical to the pre-existing `n_estimators` value. This means that the above line overwrites the `n_estimators` value - with the exact same value. Removing that if statement had no effect on reproducibility.

I think that any errors in other errors (such as XGBoost and RandomForest) which are caused by deleting the above code should cause my already-written reproducibility unit tests failing. However, none currently fail, so I don't think I've caused any errors. However, if you think I've missed something, or have a better suggestion for how to fix this bug, please let me know!

Reply:
Thank you so much for all the amazing details, @dannycg1996! `best_iteration_` only works when early_stopping is enabled. https://lightgbm.readthedocs.io/en/latest/pythonapi/lightgbm.LGBMClassifier.html#lightgbm.LGBMClassifier.best_iteration_ 

As I understand, best_iteration_  should default to n_estimator instead of 0. I'm not sure why it's 0. Looks like early_stopping is not working correctly here. Without early_stopping, we should not set n_estimators to best_iteration+1. 

What about updating the if statement to `if best_iteration is not None and best_iteration>1`?

Reply:
No worries!

I went for `if best_iteration is not None and best_iteration > 0` in the end - I think it's fine. I did consider simplifying this to just `if best_iteration` as both None and 0 are falsy, but looking at the LGBM codebase, in some places `best_iteration=-1` when initialised, and that would cause a bug!
I'll admit that I'm sceptical about whether this whole `best_iteration` code is actually serving much of a useful purpose for any estimators, but I understand your hesitation in making major tweaks that could have unintended consequences.

I think the issue with best_iteration being returned as 0 might be similar to [this](https://stackoverflow.com/questions/75867049/lgbms-best-iteration-is-none-when-using-early-early-stopping-callback-even-tho) StackOverflow post by the way. Basically with certain settings the LGBM best_iteration_ parameter doesn't work as it should. If I can make an easily reproducible example of this, I may post an issue on the LGBM GitHub at some point.

I also added another (minor change) to the codebase here, as it's been annoying me since I started working on LGBMs! Basically I've merged two if statements i.e. from this:

![image](https://github.com/user-attachments/assets/d61159a7-4e46-4c3b-90f0-ac0610146cc1)

to this:
![image](https://github.com/user-attachments/assets/2efc1c4f-67d9-45b5-8728-8b03f15b995c)

I hope that's OK - it seems a sensible minor refactor to me!

Let me know what you think - thanks!